source/common/stats/tag_utility.h

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include "envoy/stats/symbol_table.h"
+#include "envoy/stats/tag.h"
+
+#include "common/stats/symbol_table_impl.h"
+
+namespace Envoy {
+namespace Stats {
+namespace TagUtility {
+
+/**
+ * Combines a stat name with its tag to create the final stat name to use. The resulting
+ * StatNames will be valid through the lifetime of this object and all provided stat names.
+ */
+class TagStatNameJoiner {
+public:
+  /**
+   * Combines a prefix, stat name and tags into a single stat name.
+   * @param prefix StaName the stat prefix to use.
+   * @param name StaName the stat name to use.
+   * @param stat_name_tags StatNameTagVector the stat name tags to add to the stat name.
+   */
+  TagStatNameJoiner(StatName prefix, StatName stat_name, const StatNameTagVector& stat_name_tags,
+                    SymbolTable& symbol_table);
+
+  /**
+   * Combines a stat name and tags into a single stat name.
+   * @param name StaName the stat name to use.
+   * @param stat_name_tags StatNameTagVector the stat name tags to add to the stat name.
+   */
+  TagStatNameJoiner(StatName stat_name, const StatNameTagVector& stat_name_tags,
+                    SymbolTable& symbol_table);
+
+  /**
+   * @return StatName the full stat name.
+   */
+  StatName fullStatName() const { return StatName(full_name_storage_.get()); }
+
+  /**
+   * @return StatName the stat name without the tags appended.
+   */
+  StatName statNameNoTags() const { return name_; }
```

Comment:
Note that this is elsewhere called `tagExtractedStatName()`, though that's due to the process of starting with the stat-names including embedded tags, and then extracting the tags through scary slow regexes :)

In this case the tags weren't extracted; they weren't there in the first place. But maybe we should have a TODO to rename the other uses and definitions of tagExtractedStatName to use this clearer name.

I guess you could also, for symmetry, rename `fullStatName()` to `statNameWithTags()`, and change analogous definitions and uses of `statName()` to that. That would be slightly more annoying as there are many existing `statName()` methods that are agnostic to tags, and shouldn't be changed, such as the ones in symbol_table_impl.h.

Food for thought, and I'll leave it up to you and the @envoyproxy/senior-maintainers .

Reply:
I was thinking about that and ended up not wanting to go through with that refactor due to the nuances you mentioned. I figured we could think of the tag extracted name as "there are no embedded tags", so it kinda works even now. I think I'll update this site to use tagExtractedName and leave a TODO to address this later